DL full event reconstruction

TjarkMiener opened the pull request:
This PR includes three major features:

1. **Energy and arrival direction regression** (Impact parameter and shower maximum opitonal)

    - The regression tasks have been added to CTLearn. Full event reconstruction can be performed in both modes (Singletask Learning and Multitask Learning). However, in the MTL approach the protons are not excluded in the loss calculation and the task weights are fix constant from the config file. This will be solved in a follow-up PR.

2. **Residual blocks**

    - Residual blocks and a stack function have been added to the engine (Basic). This allows the user to set up a custom ResNet architecture from the config file.
    References:
        - [Deep Residual Learning for Image Recognition](https://arxiv.org/pdf/1512.03385.pdf)

3. **Attention mechanisms** 

    - Three different attention mechanisms have been added to Basic, which can be used in all four CTLearn models.
    References:
        - [Squeeze and Excitation Networks](https://arxiv.org/abs/1709.01507)
        - [Concurrent Spatial and Channel Squeeze & Excitation in Fully Convolutional Networks](https://arxiv.org/abs/1803.02579)

Some minor features:
- CTLearn can be run in 'trainandpredict' to predict after training.
- Multiple runs with the same model, but different random seeds can be performed sequentially with one command.
- Minor restructuring of the run_model.py. Separating the data loading (setting up the DL1DH: inputfn() etc.) and the output handling from the main script.
------------------------
This PR requires [DL1DH PR 85](https://github.com/cta-observatory/dl1-data-handler/pull/85).

TjarkMiener commented:
This PR would be ready for code review! Thank you, @nietootein @aribrill!

TjarkMiener commented:
Thanks @LucaRomanato for reporting a bug occur in the subfolder structure. `6733ed0` should fix this!

aribrill commented:
I am in the process of testing the PR on shevek and will have more comments soon. One issue I came across so far is the installation. I followed the installation instructions in the README. Are they still up to date, e.g. regarding installing DL1DH with pip? I had to run the following commands after setup to get things working:

```bash
source activate ctlearn
conda install -c cta-observatory ctapipe-extra
pip uninstall numpy
pip install numpy
```

aribrill commented on README.md:
````diff
@@ -120,12 +121,14 @@ Run CTLearn from the command line:
 python $CTLEARN_DIR/run_model.py myconfig.yml [--mode <MODE>] [--debug] [--log_to_file]
 ```
-`--mode <MODE>`: Set run mode with `<MODE>` as `train`, `predict`, or `load_only`. If not set, defaults to `train`.
+`--mode <MODE>`: Set run mode with `<MODE>` as `train`, `predict`, `trainandpredict`, or `load_only`. If not set, defaults to `train`.
````
```suggestion
`--mode <MODE>`: Set run mode with `<MODE>` as `train`, `predict`, `train_and_predict`, or `load_only`. If not set, defaults to `train`.
```

aribrill commented on ctlearn/default_models/basic.py:
```diff
@@ -11,26 +12,30 @@ def conv_block(inputs, training, params, reuse=None):
         bottleneck_filters = params['basic']['conv_block']['bottleneck']
         batchnorm = params['basic']['conv_block'].get('batchnorm', False)
-        
+        attention = params.get('attention', None)
```
```suggestion
        attention = params.get('attention')
```

aribrill commented on ctlearn/output_handler.py:
```diff
@@ -0,0 +1,68 @@
+    if prediction_label in list(pd.HDFStore(h5file).keys()):
+        pd.HDFStore(h5file).remove(prediction_label)
+    pd.DataFrame(data=data).to_hdf(h5file, key=prediction_label, mode='a')
+    return
```
Should something be returned here? Otherwise, there's no need to explicitly return

aribrill commented on config/example_config.yml:
```diff
@@ -251,26 +252,33 @@ Model:
-            - 'proton'
-            - 'gamma'
+    # These options have to be in consonance with the 'event_info' and 
+    # 'transforms' in 'Data'.
```
It would helpful to document what the necessary transforms and event_info are for each task.

TjarkMiener replied:
Done

aribrill commented on ctlearn/run_model.py:
```diff
@@ -506,10 +358,66 @@ def write(prediction):
         '--log_to_file',
         action='store_true',
         help="log to a file in model directory instead of terminal")
+    parser.add_argument(
```
Following an offline discussion with @TjarkMiener and @LucaRomanato, the `multiple_runs` command line argument should be removed entirely. It doesn't work well in practice, and the same functionality can be obtained if desired using `run_multiple_configurations.py`. Can you take it out (and update the Readme to match)?

TjarkMiener replied:
Done

aribrill commented on ctlearn/run_model.py:
```diff
@@ -506,10 +358,66 @@ def write(prediction):
+            config['Logging']['add_seed'] = True
+            print("CTLearn run {} with random seed '{}':".format(run+1,config['Data']['seed']))
+
+        if args.mode == 'train' or args.mode == 'load_only':
+            run_model(config, mode=args.mode, debug=args.debug, log_to_file=args.log_to_file)
+        elif args.mode == 'predict':
+            for key in config['Prediction']['prediction_file_lists']:
+                with open(args.config_file, 'r') as config_file:
+                    config = yaml.safe_load(config_file)
+                config['Data']['seed'] = random_seed
+                if args.multiple_runs != 1 or args.random_seed != 0:
+                    config['Logging']['add_seed'] = True
+                config['Data']['shuffle'] = False
+                config['Prediction']['prediction_label'] = key
+                run_model(config, mode=args.mode, debug=args.debug, log_to_file=args.log_to_file)
+        elif args.mode == 'train_and_predict':
+            run_model(config, mode='train', debug=args.debug, log_to_file=args.log_to_file)
+            for key in config['Prediction']['prediction_file_lists']:
+                with open(args.config_file, 'r') as config_file:
+                    config = yaml.safe_load(config_file)
+                config['Data']['seed'] = random_seed
+                if args.multiple_runs != 1 or args.random_seed != 0:
+                    config['Logging']['add_seed'] = True
+                config['Data']['shuffle'] = False
+                config['Prediction']['prediction_label'] = key
+                run_model(config, mode='predict', debug=args.debug, log_to_file=args.log_to_file)
```
```suggestion
        if args.mode in ['train', 'train_and_predict', 'load_only']:
            run_model(config, mode=args.mode, debug=args.debug, log_to_file=args.log_to_file)
        if args.mode in ['predict', 'train_and_predict']:
            for key in config['Prediction']['prediction_file_lists']:
                with open(args.config_file, 'r') as config_file:
                    config = yaml.safe_load(config_file)
                config['Data']['seed'] = random_seed
                if args.multiple_runs != 1 or args.random_seed != 0:
                    config['Logging']['add_seed'] = True
                config['Data']['shuffle'] = False
                config['Prediction']['prediction_label'] = key
                run_model(config, mode=args.mode, debug=args.debug, log_to_file=args.log_to_file)
```

TjarkMiener replied:
Thanks for the suggestion! I sightly modified it to make it work. The run_model function not have a 'train_and_predict' mode, so we have to call it recursively and manually.

I search for "train" and "predict" in the args.mode string. So it will also work accordingly if the user set the mode to "training", "predicting", "prediction", "train_and_predict", "training_and_predicting" ...
The mode "load_only" is separated.

What do you think?

aribrill replied:
Clever! Seems good to me.

aribrill commented on ctlearn/utils.py:
```diff
@@ -0,0 +1,83 @@
+import importlib
+import logging
+import numpy as np
+import os
+import pkg_resources
+import sys
+import pandas as pd
+import time
+import yaml
```
```suggestion
import importlib
import logging
import os
import pkg_resources
import sys
import time

import numpy as np
import pandas as pd
import yaml
```

aribrill commented on ctlearn/data_loader.py:
```diff
@@ -0,0 +1,147 @@
+import importlib
+import logging
+import numpy as np
+import os
+import pkg_resources
+import sys
+import pandas as pd
+import time
+import yaml
+
+import tensorflow as tf
```
```suggestion
import importlib
import logging
import os
import pkg_resources
import sys
import time

import numpy as np
import pandas as pd
import tensorflow as tf
import yaml
```

aribrill reviewed (changes requested):
The PR is very nice! I am currently replicating one of the new benchmarks as a test.

Some other comments:
- Please edit the Readme to bring it up to date with the PR, e.g. update the installation instructions with all required dependencies and specified DL1DH version; remove references to deprecated features like multi-stereo mode; and make sure that the command line options listed in the "Run a Model" section are the right ones.
- In utils.py, compute_class_weights() and log_examples() aren't compatible with the new multitask config format. I can commit fixes tomorrow after the benchmark test is done.
- Computing the number of examples when a label is a sequence such as (alt, az) also required a change to DL1DH, which I have requested as cta-observatory/dl1-data-handler#99.

TjarkMiener commented:
Thanks for the comments @aribrill! I resolved most of them. Regarding the installation, can you specify the ctapipe and dl1dh version you used please? ctapipe-extra should not be required though.

Also thanks for fixing the number of examples of DL1DH. I will have a look into it now.

aribrill commented:
I used DL1DH v0.8.2, which requires ctapipe 0.8.0.

On Tue, Mar 2, 2021 at 9:54 AM Tjark Miener <notifications@github.com>
wrote:

> Thanks for the comments @aribrill <https://github.com/aribrill>! I
> resolved most of them. Regarding the installation, can you specify the
> ctapipe and dl1dh version you used please? ctapipe-extra should not be
> required though.
>
> Also thanks for fixing the number of examples of DL1DH. I will have a look
> into it now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ctlearn-project/ctlearn/pull/132#issuecomment-788965481>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEWJLHIWMHUX3XEZGAMPTR3TBT3ZBANCNFSM4PEPLJWA>
> .
>

TjarkMiener commented:
The correct versions of DL1DH and ctapipe should be installed automatically by conda, when the environment is created. There is no need to git clone the DL1DH repo. The latest commits should fix the installation and I updated the README. Can you please check, if the correct versions are installed on `shevek`? I also added a commit to make ctlearn a command line tool to avoid pointing to the run_model script. This should work: `$ ctlearn myconfig.yml`

aribrill reviewed (changes requested):
Looking great! I can see only one small additional change to make resulting from the recent revisions.

Thanks for updating the installation method, I can confirm that it works on shevek (although the compatibility PR is not included). I really like the ability to use ctlearn as a command line tool.

aribrill commented:
Actually, I have a question. I replicated one of the benchmark models (arr_lowcut_TRN) and would like to plot the predictions for diffuse gammas. I have a predictions file in h5 format. I see that none of the plotting scripts have been updated to be compatible with this new format, however. Are there new scripts that replicate the old functionality or at least some instructions for what to do with this file?

I also see that ctaplot has been added as a dependency, but I can't find where it's used. Is that related in some way?

TjarkMiener commented on ctlearn/run_model.py:
```diff
@@ -2,103 +2,44 @@
+        os.makedirs(root_model_dir)
+
+    random_seed = config['Data']['seed']
+    model_dir += "/seed_{}".format(random_seed)
```
Done!

TjarkMiener commented:
Thanks for the fixes! Everything worked fine for the compatibility PR. I uploaded the visualisation notebook. Anything missing?

aribrill reviewed (approved):
Looks good! I updated the particle classification plotting scripts for compatibility with the new predictions output format.

TjarkMiener commented:
Thanks for updating the plotting scripts!
